airflow/__init__.py

```diff
@@ -81,6 +82,12 @@
     # Deprecated lazy imports
     "AirflowException": (".exceptions", "AirflowException", True),
 }
+if TYPE_CHECKING:
+    # This is never executed, but tricks static analyzers (PyDev, PyCharm,)
+    # into knowing the types of these symbols, and what they contain.
+    from airflow.models.dag import DAG  # noqa: TCH004
+    from airflow.models.dataset import Dataset  # noqa: TCH004
+    from airflow.models.xcom_arg import XComArg  # noqa: TCH004
```

Comment:
I think instead of doing `noqa` you can add these to `__all__` instead. Not sure. That would be a lot better if it works.

Reply:
There are already in `__all__`, as result ruff complain that it should be moved outside of TYPE_CHECKING (TCH004) block because it unable to detect PEP 562 imports.

Reply:
I thought as an option to add this rule into the in exclusion list into the `pyproject.toml`, by the same way as we already done for `airflow/models/__init__.py`

https://github.com/apache/airflow/blob/242a169a81b203b3618c312d8a2ec9b4c8efd121/pyproject.toml#L346-L348

Reply:
Already: https://github.com/apache/airflow/pull/39750

Reply:
As soon as we merge https://github.com/apache/airflow/pull/39750 I will drop commit which add this fix from this pr